tracer/src/Datadog.Trace/Debugger/Instrumentation/MethodDebuggerInvoker.MultiProbe.cs

```diff
@@ -0,0 +1,197 @@
+
+namespace Datadog.Trace.Debugger.Instrumentation
+{
+    /// <summary>
+    /// MethodDebuggerInvoker for multiple probes scenario (where there are more than one _method_ probe).
+    /// </summary>
+    public static partial class MethodDebuggerInvoker
+    {
+        /// <summary>
+        /// Begin Method Invoker that accepts multiple probes
+        /// </summary>
+        /// <typeparam name="TTarget">Target type</typeparam>
+        /// <param name="instance">Instance value</param>
+        /// <param name="methodMetadataIndex">The index used to lookup for the <see cref="MethodMetadataInfo"/> associated with the executing method</param>
+        /// <param name="instrumentationSequence">The version of the instrumentation</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static MethodDebuggerState[] BeginMethod_StartMarker<TTarget>(TTarget instance, int methodMetadataIndex, int instrumentationSequence)
+        {
+            if (!MethodMetadataCollection.Instance.IndexExists(methodMetadataIndex))
+            {
+                Log.Warning("BeginMethod_StartMarker: Failed to receive the InstrumentedMethodInfo associated with the executing method. type = {Type}, instance type name = {Name}, methodMetadaId = {MethodMetadataIndex}, instrumentationSequence = {InstrumentationSequence}", new object[] { typeof(TTarget), instance?.GetType().Name, methodMetadataIndex, instrumentationSequence });
+                return CreateInvalidatedDebuggerStates();
+            }
+
+            ref var methodMetadataInfo = ref MethodMetadataCollection.Instance.Get(methodMetadataIndex);
+            var probesIds = methodMetadataInfo.ProbeIds;
+            var probeMetadataIndices = methodMetadataInfo.ProbeMetadataIndices;
+
+            if (instrumentationSequence != methodMetadataInfo.InstrumentationSequence)
+            {
+                Log.Warning("BeginMethod_StartMarker: The instrumentation sequence is different, received sequence: {InstrumentationSequence}, but the kept sequence is: {KeptSequenceNumber}", new object[] { instrumentationSequence, methodMetadataInfo.InstrumentationSequence });
+                return CreateInvalidatedDebuggerStates();
+            }
+
+            var states = InstrumentationAllocator.RentArray<MethodDebuggerState>(probesIds.Length);
+
+            for (var stateIndex = 0; stateIndex < states.Length; stateIndex++)
+            {
+                states[stateIndex] = BeginMethod_StartMarker(instance, methodMetadataIndex, probeMetadataIndices[stateIndex], probesIds[stateIndex]);
+            }
+
+            return states;
+        }
+
+        /// <summary>
+        /// Ends the markering of BeginMethod.
+        /// </summary>
+        /// <param name="states">Debugger state</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void BeginMethod_EndMarker(ref MethodDebuggerState[] states)
```

Comment:
I think the name `MethodDebuggerState` was already a bit vague, but with the changes in this PR, I think it's now a misleading/incorrect, since we're using it here to represent that state of each probe within a particular method execution. Could we come up with a different name, perhaps `MethodProbeState` or `MethodProbeEvaluationState`?

Reply:
I agree with your sentiment. Having said that, I prefer doing such naming refactoring in a separate dedicated PR to avoid confusing code reviewing and future auditing.
In my subsequent PR I plan to do some refactoring and cleanups in our instrumentation logic and consolidate a lot so I'll do that there.
WDYT?

Reply:
Sounds good.